File: services/operator_ui/src/views/__tests__/runModel.test.ts
import { describe, expect, it } from "vitest";
import type { TraceSummary } from "../../api/types";
import {
  filterRuns,
  runBucketCounts,
  runOutcomeLabel,
  runOutcomeTone,
} from "../runModel";

describe("operator run model", () => {
  it("separates active, attention, successful, failed, and hidden runs", () => {
    const rows = [
      trace("ACTIVE", "in-flight", "In Flight"),
      trace("STUCK", "stuck", "Stuck"),
      trace("DONE", "done", "Complete"),
      trace("FAIL", "done", "Failed"),
      { ...trace("HIDE", "hidden", "Complete"), hidden: true },
    ];

    expect(runBucketCounts(rows)).toEqual({
      active: 1,
      attention: 2,
      recent: 4,
      successful: 1,
      failed: 1,
      hidden: 1,
    });
    expect(filterRuns(rows, "attention").map((row) => row.id)).toEqual([
      "STUCK",
      "FAIL",
    ]);
    expect(filterRuns(rows, "successful").map((row) => row.id)).toEqual(["DONE"]);
  });

  it("surfaces submitted multi-day runs as attention", () => {
    const row = { ...trace("OLD", "done", "Submitted"), elapsed: ">24h (multi-run)" };

    expect(filterRuns([row], "attention")).toHaveLength(1);
    expect(runOutcomeLabel(row)).toBe("Needs attention");
    expect(runOutcomeTone(row)).toBe("warn");
  });
});

function trace(id: string, status: TraceSummary["status"], rawStatus: string): TraceSummary {
  return {
    id,
    title: "Test run",
    status,
    raw_status: rawStatus,
    hidden: false,
    lifecycle_state: "",
    state_reason: "",
    run_id: `trace-${id}`,
    phase: "",
    elapsed: "1m",
    started_at: "2026-05-04T12:00:00+00:00",
    pr_url: null,
    pipeline_mode: "",
    review_verdict: "",
    qa_result: "",
  };
}

File: services/operator_go/internal/dashboard/run_model_test.go
package dashboard

import "testing"

func TestRunBucketCountsMatchOperatorModel(t *testing.T) {
	rows := []TraceSummary{
		{ID: "A", Status: "in-flight", RawStatus: "Running"},
		{ID: "B", Status: "stuck", RawStatus: "Blocked"},
		{ID: "C", Status: "done", RawStatus: "Complete"},
		{ID: "D", Status: "done", RawStatus: "Failed"},
		{ID: "E", Status: "done", RawStatus: "Submitted", Elapsed: ">24h"},
		{ID: "F", Status: "hidden", RawStatus: "Suppressed", Hidden: true},
	}

	counts := RunBucketCounts(rows)

	if counts[BucketActive] != 1 {
		t.Fatalf("active = %d, want 1", counts[BucketActive])
	}
	if counts[BucketAttention] != 3 {
		t.Fatalf("attention = %d, want 3", counts[BucketAttention])
	}
	if counts[BucketSuccessful] != 1 {
		t.Fatalf("successful = %d, want 1", counts[BucketSuccessful])
	}
	if counts[BucketFailed] != 1 {
		t.Fatalf("failed = %d, want 1", counts[BucketFailed])
	}
	if counts[BucketHidden] != 1 {
		t.Fatalf("hidden = %d, want 1", counts[BucketHidden])
	}
}

func TestRunOutcomePrioritizesFailureBeforeAttention(t *testing.T) {
	row := TraceSummary{ID: "FAIL-1", Status: "done", RawStatus: "Escalated", LifecycleState: "stale"}

	if got := RunOutcomeLabel(row); got != "Failed" {
		t.Fatalf("RunOutcomeLabel() = %q, want Failed", got)
	}
}

func TestFilterRunsExcludesHiddenFromRecent(t *testing.T) {
	rows := []TraceSummary{
		{ID: "A", Status: "done", RawStatus: "Complete"},
		{ID: "B", Status: "hidden", RawStatus: "Suppressed", Hidden: true},
	}

	recent := FilterRuns(rows, BucketRecent)
	if len(recent) != 1 || recent[0].ID != "A" {
		t.Fatalf("recent = %#v, want only A", recent)
	}
}
